docs/design/otlp.md

```diff
@@ -0,0 +1,171 @@
+| Reqwest (async) | `reqwest-client` | No (Tokio) | `reqwest` | HTTP/1.1 → auto-upgrades to HTTP/2 over TLS | Services that already run a Tokio runtime and prefer fully non-blocking exports.
+| Hyper | `hyper-client` | No (Tokio) | `hyper` | HTTP/1.1 → auto-upgrades to HTTP/2 over TLS | Lean dependency footprint when both Reqwest features are disabled.
+| Tonic (gRPC) | `grpc-tonic` | No (Tokio) | `tonic` | HTTP/2 (gRPC) | Environments standardising on gRPC or collectors exposing only port 4317.
+| Custom | — | Depends | user-supplied | depends on implementation | Embedding a bespoke `HttpClient` via `.with_http_client()`.
+
+---
+
+## 5. Runtime Flow (Trace Export)
+```mermaid
+sequenceDiagram
+    participant App as Application code
+    participant SDK as opentelemetry-sdk
+    participant Exp as SpanExporter
+    participant Tx as Transport Client (HTTP/Reqwest or gRPC/Tonic)
+    participant Collector as OTLP Collector
+    App->>SDK: span.start / span.end
+    SDK-->>SDK: BatchSpanProcessor collects SpanData
+    SDK->>Exp: export(batch)
+    Exp->>Tx: serialize + send
+    Tx--)Collector: OTLP request
+    note right of Collector: 2xx = success,<br/>non-2xx → error handling
+    Tx--)Exp: Future resolves
+    Exp--)SDK: OTelSdkResult
+```
+Notes: 
+
+* Serialization happens **inside the transport client** module to keep exporters thin.
+* `export` is `async` for `tonic` / `reqwest` clients, surfaced via `SpanExporter` implementing `opentelemetry_sdk::trace::SpanExporter`.
+* Resource attributes are injected once per exporter via `set_resource()` before first export.
+
+---
+
+## 6. Configuration & Environment Variable Resolution
+`exporter::mod.rs` implements helper fns:
+* `default_protocol()` – chosen from compile-time defaults.
+* `resolve_timeout()` – precedence: signal-specific env → generic env → builder → default 10 s.
+* `parse_header_string()` – parses comma-separated `k=v` pairs with URL-decoding.
+
+Signal builders read **signal-specific env vars** (e.g. `OTEL_EXPORTER_OTLP_TRACES_ENDPOINT`) _before_ generic ones, matching the spec.
+
+---
+
+## 7. Error Handling Strategy
+`ExporterBuildError` is a non-exhaustive enum covering:
+* Builder-time validation (URI parse, missing client)
+* Feature gating errors (compression requested w/o feature)
+* Runtime errors are wrapped in `OTelSdkError` from the SDK.
+
+Builder fails fast, runtime exporter surfaces errors through `export()` futures so processors (or Retry logic) can decide whether to back-off, drop, or escalate.
+
+---
+
+## 8. Extension & Customisation Points
+1. **Custom headers / metadata** – `.with_metadata(map)` (gRPC) or `.with_headers()` (HTTP).
+2. **Compression** – `.with_compression(Compression::Gzip | Compression::Zstd)` behind feature-flags.
+3. **TLS** – via `TonicConfig` or `HttpConfig`; root-store helper features embed common CA bundles.
+4. **Alternate HTTP client** – inject any `HttpClient` implementation via `.with_http_client()`.
```

Comment:
this is something missing from docs I think! Do we actually support bring-your-own-client capability!

Reply:
The code's there and it's a public trait and method so it should work!